065-topic-replication-factor-change.md

```diff
@@ -0,0 +1,290 @@
+This is not really an option for most use cases.
+
+## Motivation
+
+The topic replication factor change feature will provide a much better cloud-native experience for Strimzi users.
+They will be able to change the topic replication factor by simply updating the KafkaTopic resource, which may also happen through a GitOps pipeline.
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+A tighter integration with Cruise Control is also one of the long term goals of the Strimzi project. 
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.
+
+Cruise Control creates a task for each operation, such as topic configuration and cluster rebalance.
+A task is first created in the Active state, then moves to InExecution, and finally to Completed or CompletedWithError.
+At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
+During the execution, a task is divided in one or more batches, which are then executed one after the other.
+A task may include multiple topics and it is marked as completed once the whole task's status is Completed.
+
+### Topic configuration
+
+The [`topic_configuration`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#change-kafka-topic-configuration) endpoint is used to request a topic configuration update, and currently only supports replication factor changes.
+Once accepted, the replication factor change is put into the Cruise Control's task queue and executed asynchronously.
+
+The two required parameters for this endpoint are `topic` (regular expression to specify subject topics) and `replication_factor` (target replication factor).
+In order to group multiple changes in the same request to Cruise Control, the Topic Operator uses a [JSON payload](https://github.com/linkedin/cruise-control/wiki/Change-topic-replication-factor-through-Cruise-Control#instruction), which replaces the required parameters.
+This JSON contains a list of replication factor values, each one associated with a regex expression of the form `topic1|topic2|topic3` matching all topics with that target replication factor.
+
+This is useful with bulk replicas change, which can happen in some edge cases.
+A potentially big backlog of messages may be created when the Topic Operator is not running, or there is an issue with Cruise Control or Kafka.
```

Comment:
So in this situation, the TO would make a single topic_config request to the CC API with the JSON payload specifying all the topics to be configured? 

Assuming users are making topic replication factor changes through the `KafkaTopic` resources, is the TO batching these changes during a reconciliation loop before making the request? If so, how would the batching happen on the TO side? Would the batching be based on time or number of topics, would it be configurable by users?

 Is there a way users could batch these changes on purpose?

Reply:
> So in this situation, the TO would make a single topic_config request to the CC API with the JSON payload specifying all the topics to be configured?

Yes.

> is the TO batching these changes during a reconciliation loop before making the request? If so, how would the batching happen on the TO side? Would the batching be based on time or number of topics, would it be configurable by users?

The batching logic is already provided by the reconciliation logic. On each iteration, it reconciles a batch of KafkaTopic events, some of which may be replicas changes, and you can tune the batch.size and linger.ms.

> Is there a way users could batch these changes on purpose?

What the user can do is to accumulate all replicas changes, and then apply them at the same time. This would produce 1 or more batches depending on batch.size and linger.ms configurations. I tried 100 replicas changes with default configurations, and it created 3 batches. This means 3 CC topic_configuration requests, plus a small number of tasks_state requests (it took something like 5 minutes to reconcile all).